crawler/urlJobProcessor.js

```diff
@@ -1,36 +1,106 @@
 
 const probes = [];
 
-function configure(/* opts */) {
-  // TODO: Set up probes. They'll be specified somehow in opts and we
-  // should be able to load each probe individually. Maybe something
-  // like `require('./probe-' + probeName + '.js');`
+function configure(opts) {
+  if (!opts.probes || !opts.probes.length) {
+    return Promise.reject(new Error('At least one urlJobProcessor probe must be specified'));
+  }
+
+  const ret = [Promise.resolve()];
+
+  opts.probes.forEach((probeOpts) => {
+    if (!probeOpts.name) {
+      return Promise.reject(new Error('Probe without name encountered'));
+    }
+
+    // We allow the probe to be specified by name or by path. If specified by name,
+    // we expect it to be a known subdir.
+    const path = probeOpts.path || `./urlJobProcessorProbes/${probeOpts.name}.js`;
+    const probeModule = require(path).default;
+
+    if (!probeModule.runProbe) {
+      return Promise.reject(new Error(`Probe ${probeOpts.name} does not implement runProbe`));
+    }
+
+    let p = Promise.resolve();
+    if (probeModule.configure) {
+      p = probeModule.configure(probeOpts.opts || {});
+    }
+
+    ret.push(p.then(() => {
+      probes.push({
+        name: probeOpts.name,
+        probeFn: probeModule.runProbe,
+      });
+    }));
+  });
+
+  return Promise.all(ret);
 }
 
-function fetchAllResources(url) {
-  const ret = { html: '', js: '', sw: '' };
-  return fetch(`http://${url}`).then((response) => {
-    return response.text().then((htmlText) => {
-      ret.html += htmlText;
+function fetchAllResources(urlArg) {
+  const url = urlModule.parse(`http://${urlArg}`);
+
+  // TODO: Pass user agent in headers
+  return fetch(`${url.href}`, { headers: '' }).then((mainResponse) => {
+    // We pass along the response object that was returned after
+    // our fetch of the main page. As an optimization, we don't
+    // clone the response here. Instead, we pass the body text
+    // in a separate arg.
+    const responses = [mainResponse];
+    return mainResponse.text().then((htmlText) => {
+      const $ = cheerio.load(htmlText);
+      const scripts = [];
+      const promises = [];
+
+      $('script').each((index, elem) => {
+        if (!$(elem).attr('src')) {
+          // For inline scripts, just append them to the array of scripts
+          // that we'll pass to the probes
+          scripts.push($(elem).text());
+        } else {
+          // For scripts with a `src` attribute, we must fetch them here
+          let src = $(elem).attr('src');
+          if (src[0] === '/') {
+            // Deal with relative paths in `src` attributes
+            src = `${url.protocol}//${url.host}${src}`;
+          }
+          promises.push(fetch(src).then((scriptResponse) => {
+            return scriptResponse.text();
+          }).then((jsText) => {
+            return scripts.push(jsText);
+          }));
+        }
+      });
+
+      return Promise.all(promises).then(() => {
```

Comment:
> `Promise.all` will fail if only one fetch fails. Can be fixed by adding a `catch` to each fetch chain.

This was intentional. If we fail to fetch one of the scripts that a page uses then we might not find (for example) a Service Worker when there actually is one. We make the whole job fail because that means we'll try again later. If we ignore fetch failures then we won't know that something went wrong for a particular data point.